SourceNode.ForceStop can nil-panic if force-stop races source-node startup (invariant 7)

## Summary

`SourceNode.ForceStop` (`pkg/lifecycle/stream/source.go:248-253`) dereferences `n.connectorCtxCancel` unconditionally:

```go
func (n *SourceNode) ForceStop(ctx context.Context) {
	n.logger.Warn(ctx).Msg("force stopping source connector")
	n.mctx.Lock()
	n.connectorCtxCancel()   // <-- nil panic if Run hasn't set it yet
	n.mctx.Unlock()
}
```

`connectorCtxCancel` is assigned at `source.go:110`, **inside** `SourceNode.Run` (which executes in a tomb goroutine). The pipeline status is set to `StatusRunning` at `service.go:774` in a *different* goroutine. There is no ordering between the two, so there is a window where `Service.Stop(ctx, id, force=true)` passes its `StatusRunning` guard and calls `node.ForceStop`, while the source node's `Run` has not yet reached line 110 → `n.connectorCtxCancel` is nil → **panic (SIGSEGV)**.

## How it was found

Surfaced by a new deterministic regression test for #2521 that force-stopped a pipeline with zero startup delay. Reproduced as:

```
panic: runtime error: invalid memory address or nil pointer dereference
  pkg/lifecycle/stream/source.go:251  (*SourceNode).ForceStop
  pkg/lifecycle/service.go:317        (*Service).stopForceful
```

The existing forceful-stop test masks it with a `time.Sleep(100ms)` before force-stopping; under CI load that margin is not guaranteed.

## Why it matters

Data-integrity invariant 7: *"kill -9 at any instant must be recoverable."* Force-stop is our explicit "kill it now" path; it must not itself crash the process when invoked during the start/recover window. A panic here takes down the engine goroutine rather than degrading the pipeline.

## Proposed fix (needs its own analysis — do not just nil-guard blindly)

The naive fix is a nil check, but `connectorCtxCancel` is written in `Run` and read in `ForceStop` — confirm both happen under `n.mctx` (the write at `source.go:110` should be inside the same lock as the read) so the guard isn't itself racy. Options:
1. Assign `connectorCtxCancel` under `n.mctx` in `Run`, and nil-guard the read in `ForceStop`. If it's still nil (Run not started), the tomb kill in `stopForceful` (which fires *before* the `ForceStop` loop) already cancels the shared node context, so skipping the connector-ctx cancel is safe — the node stops when its `Run` starts and observes the dead tomb.
2. Initialize `connectorCtxCancel` to a no-op cancel at construction so it's never nil.

Either way this is a Tier-1 lifecycle concurrency change and should ship with a test that force-stops during the startup window under `-race -count`.

Related: #2521 (the WaitPipeline fix that surfaced this), #1659 (force-stop / arch semantics), `docs/design-documents/20260706-forceful-stop-test-determinism.md`.

--- a/pkg/lifecycle/stream/destination.go
+++ b/pkg/lifecycle/stream/destination.go
@@ -36,7 +36,9 @@ type DestinationNode struct {
 	base   pubSubNodeBase
 	logger log.CtxLogger
 
-	connectorCtxCancel context.CancelFunc
+	// stopper force-stops the connector context, coordinating with Run so a
+	// force-stop that races startup is neither lost nor a nil-panic. See #2539.
+	stopper forceStopper
 }
 
 type Destination interface {
@@ -63,10 +65,10 @@ func (n *DestinationNode) Run(ctx context.Context) (err error) {
 	defer cleanup()
 
 	// start a fresh connector context to make sure the connector is running
-	// until this method returns
-	var connectorCtx context.Context
-	connectorCtx, n.connectorCtxCancel = context.WithCancel(context.Background())
-	defer n.connectorCtxCancel()
+	// until this method returns. stopper.start also applies a force-stop that
+	// arrived before now (raced startup).
+	connectorCtx, connectorCtxCancel := n.stopper.start()
+	defer connectorCtxCancel()
 
 	var (
 		// lastPosition stores the position of the last successfully processed record
@@ -137,7 +139,7 @@ func (n *DestinationNode) Run(ctx context.Context) (err error) {
 
 func (n *DestinationNode) ForceStop(ctx context.Context) {
 	n.logger.Warn(ctx).Msg("force stopping destination connector")
-	n.connectorCtxCancel()
+	n.stopper.stop()
 }
 
 // Sub will subscribe this node to an incoming channel.

--- a/pkg/lifecycle/stream/destination_acker.go
+++ b/pkg/lifecycle/stream/destination_acker.go
@@ -36,13 +36,12 @@ type DestinationAckerNode struct {
 
 	queueMutex sync.Mutex
 
-	// mctx guards access to the contextCtxCancel function
-	mctx sync.Mutex
-
 	base   subNodeBase
 	logger log.CtxLogger
 
-	connectorCtxCancel context.CancelFunc
+	// stopper force-stops the connector context, coordinating with Run so a
+	// force-stop that races startup is neither lost nor a nil-panic. See #2539.
+	stopper forceStopper
 }
 
 func (n *DestinationAckerNode) ID() string {
@@ -51,12 +50,10 @@ func (n *DestinationAckerNode) ID() string {
 
 func (n *DestinationAckerNode) Run(ctx context.Context) (err error) {
 	// start a fresh connector context to make sure the connector is running
-	// until this method returns
-	var connectorCtx context.Context
-	n.mctx.Lock()
-	connectorCtx, n.connectorCtxCancel = context.WithCancel(context.Background())
-	n.mctx.Unlock()
-	defer n.connectorCtxCancel()
+	// until this method returns. stopper.start also applies a force-stop that
+	// arrived before now (raced startup).
+	connectorCtx, connectorCtxCancel := n.stopper.start()
+	defer connectorCtxCancel()
 
 	// signalChan is buffered to ensure signals don't get lost if worker is busy
 	signalChan := make(chan struct{}, 1)
@@ -253,7 +250,5 @@ func (n *DestinationAckerNode) SetLogger(logger log.CtxLogger) {
 
 func (n *DestinationAckerNode) ForceStop(ctx context.Context) {
 	n.logger.Warn(ctx).Msg("force stopping destination acker node")
-	n.mctx.Lock()
-	n.connectorCtxCancel()
-	n.mctx.Unlock()
+	n.stopper.stop()
 }

--- a/pkg/lifecycle/stream/dlq.go
+++ b/pkg/lifecycle/stream/dlq.go
@@ -59,7 +59,9 @@ type DLQHandlerNode struct {
 	// m guards access to Handler and window
 	m sync.Mutex
 
-	handlerCtxCancel context.CancelFunc
+	// stopper force-stops the handler context, coordinating with Run so a
+	// force-stop that races startup is neither lost nor a nil-panic. See #2539.
+	stopper forceStopper
 }
 
 func (n *DLQHandlerNode) ID() string {
@@ -73,10 +75,10 @@ func (n *DLQHandlerNode) Run(ctx context.Context) error {
 	n.window = newDLQWindow(n.WindowSize, n.WindowNackThreshold)
 
 	// start a fresh connector context to make sure the handler is running until
-	// this method returns
-	var handlerCtx context.Context
-	handlerCtx, n.handlerCtxCancel = context.WithCancel(context.Background())
-	defer n.handlerCtxCancel()
+	// this method returns. stopper.start also applies a force-stop that arrived
+	// before now (raced startup).
+	handlerCtx, handlerCtxCancel := n.stopper.start()
+	defer handlerCtxCancel()
 
 	err := n.Handler.Open(handlerCtx)
 	if err != nil {
@@ -207,7 +209,7 @@ func (n *DLQHandlerNode) SetLogger(logger log.CtxLogger) {
 
 func (n *DLQHandlerNode) ForceStop(ctx context.Context) {
 	n.logger.Warn(ctx).Msg("force stopping DLQ handler node")
-	n.handlerCtxCancel()
+	n.stopper.stop()
 }
 
 // dlqWindow is responsible for tracking the last N nacks/acks and enforce a

--- a/pkg/lifecycle/stream/force_stop.go
+++ b/pkg/lifecycle/stream/force_stop.go
@@ -39,6 +39,10 @@ import (
 // invariant 7: force at any instant must be recoverable, not crash), and it was
 // previously reimplemented per node — one of four correctly, the rest with a
 // nil-deref panic and, for two, an unsynchronized field. See #2539.
+//
+// A forceStopper is single-shot: once stopped it stays stopped, matching a node's
+// single-use lifecycle (a new pipeline run builds fresh nodes). It is not designed
+// to be reset and reused.
 type forceStopper struct {
 	mu      sync.Mutex
 	cancel  context.CancelFunc

--- a/pkg/lifecycle/stream/force_stop.go
+++ b/pkg/lifecycle/stream/force_stop.go
@@ -1,0 +1,74 @@
+// Copyright © 2026 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stream
+
+import (
+	"context"
+	"sync"
+)
+
+// forceStopper coordinates force-stopping a node's connector context with the
+// node's Run initialization. A force-stoppable node holds one, calls start() when
+// Run sets up its connector context, and calls stop() from ForceStop.
+//
+// The two can race: force-stop is gated only on the pipeline status being
+// "running", which is set as the node goroutines launch, so ForceStop may run
+// before Run has created the connector context. forceStopper makes that safe in
+// both orderings:
+//
+//   - stop() after start(): cancels the connector context immediately.
+//   - stop() before start(): latches the request; start() cancels the context the
+//     moment it is created, so the force-stop is neither lost nor a nil-panic.
+//
+// All state is guarded by mu, so the cancel func written by start() (in the Run
+// goroutine) and read by stop() (in the force-stop goroutine) never race.
+//
+// This exists because force-stop is a crash-safety guarantee (data-integrity
+// invariant 7: force at any instant must be recoverable, not crash), and it was
+// previously reimplemented per node — one of four correctly, the rest with a
+// nil-deref panic and, for two, an unsynchronized field. See #2539.
+type forceStopper struct {
+	mu      sync.Mutex
+	cancel  context.CancelFunc
+	stopped bool
+}
+
+// start returns a fresh cancelable context for the node's connector and records
+// its cancel func. If stop() already ran, the returned context is already
+// canceled. The caller must defer the returned cancel func so the connector
+// context is released when Run returns.
+func (f *forceStopper) start() (context.Context, context.CancelFunc) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	f.cancel = cancel
+	if f.stopped {
+		cancel()
+	}
+	return ctx, cancel
+}
+
+// stop cancels the connector context. If start() has not run yet, it latches the
+// request so start() cancels immediately when it does.
+func (f *forceStopper) stop() {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if f.cancel != nil {
+		f.cancel()
+		return
+	}
+	f.stopped = true
+}

--- a/pkg/lifecycle/stream/force_stop_test.go
+++ b/pkg/lifecycle/stream/force_stop_test.go
@@ -1,0 +1,98 @@
+// Copyright © 2026 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stream
+
+import (
+	"context"
+	"sync"
+	"testing"
+
+	"github.com/matryer/is"
+)
+
+func TestForceStopper_StopAfterStart(t *testing.T) {
+	is := is.New(t)
+	var fs forceStopper
+
+	ctx, cancel := fs.start()
+	defer cancel()
+	is.NoErr(ctx.Err()) // running: not canceled yet
+
+	fs.stop()
+	is.True(ctx.Err() != nil) // stop cancels the connector context
+}
+
+// TestForceStopper_StopBeforeStart is the #2539 case: a force-stop that arrives
+// before Run initialized the context must not be lost or panic — it latches, and
+// start() returns an already-canceled context.
+func TestForceStopper_StopBeforeStart(t *testing.T) {
+	is := is.New(t)
+	var fs forceStopper
+
+	fs.stop() // no context yet — must latch, not panic
+
+	ctx, cancel := fs.start()
+	defer cancel()
+	is.True(ctx.Err() != nil) // start applies the latched force-stop immediately
+}
+
+func TestForceStopper_StopIsIdempotent(t *testing.T) {
+	is := is.New(t)
+	var fs forceStopper
+
+	fs.stop()
+	fs.stop() // latched twice — no panic
+	ctx, cancel := fs.start()
+	defer cancel()
+	is.True(ctx.Err() != nil)
+
+	fs.stop() // after start — no panic
+	fs.stop()
+}
+
+// TestForceStopper_ConcurrentStartStop proves the invariant that matters: no
+// matter how start() (Run goroutine) and stop() (force-stop goroutine) interleave,
+// the connector context always ends canceled — the force-stop is never lost — and
+// the field access is race-free (run under -race).
+func TestForceStopper_ConcurrentStartStop(t *testing.T) {
+	is := is.New(t)
+
+	for i := 0; i < 500; i++ {
+		var fs forceStopper
+		var (
+			ctx    context.Context
+			cancel context.CancelFunc
+		)
+
+		release := make(chan struct{})
+		var wg sync.WaitGroup
+		wg.Add(2)
+		go func() {
+			defer wg.Done()
+			<-release
+			ctx, cancel = fs.start()
+		}()
+		go func() {
+			defer wg.Done()
+			<-release
+			fs.stop()
+		}()
+		close(release)
+		wg.Wait()
+
+		is.True(ctx.Err() != nil) // force-stop is honored regardless of ordering
+		cancel()
+	}
+}

--- a/pkg/lifecycle/stream/source.go
+++ b/pkg/lifecycle/stream/source.go
@@ -248,8 +248,17 @@ func (n *SourceNode) stopGraceful(ctx context.Context, reason error) (err error)
 func (n *SourceNode) ForceStop(ctx context.Context) {
 	n.logger.Warn(ctx).Msg("force stopping source connector")
 	n.mctx.Lock()
-	n.connectorCtxCancel()
-	n.mctx.Unlock()
+	defer n.mctx.Unlock()
+	// connectorCtxCancel is nil if Run has not yet reached its initialization
+	// (force-stop raced source startup). Skipping the cancel is safe: stopForceful
+	// kills the pipeline tomb before calling ForceStop, so Run will observe the
+	// canceled node context and stop as soon as it starts. Canceling the connector
+	// context here only interrupts an in-flight Read faster — which by definition
+	// isn't happening yet. Guarding avoids a nil-deref panic (invariant 7: force
+	// at any instant must be recoverable, not crash). See issue #2539.
+	if n.connectorCtxCancel != nil {
+		n.connectorCtxCancel()
+	}
 }
 
 func (n *SourceNode) Pub() <-chan *Message {

--- a/pkg/lifecycle/stream/source.go
+++ b/pkg/lifecycle/stream/source.go
@@ -44,11 +44,11 @@ type SourceNode struct {
 	base   pubNodeBase
 	logger log.CtxLogger
 
-	state              csync.ValueWatcher[nodeState]
-	connectorCtxCancel context.CancelFunc
+	state csync.ValueWatcher[nodeState]
 
-	// mctx guards access to the connector context
-	mctx sync.Mutex
+	// stopper force-stops the connector context, coordinating with Run so a
+	// force-stop that races startup is neither lost nor a nil-panic. See #2539.
+	stopper forceStopper
 
 	stop struct {
 		sync.Mutex
@@ -103,13 +103,11 @@ func (n *SourceNode) Run(ctx context.Context) (err error) {
 	defer cleanup()
 
 	// start a fresh connector context to make sure the connector is running
-	// until this method returns
-	var connectorCtx context.Context
-
-	n.mctx.Lock()
-	connectorCtx, n.connectorCtxCancel = context.WithCancel(context.Background())
-	n.mctx.Unlock()
-	defer n.connectorCtxCancel()
+	// until this method returns. stopper.start also applies a force-stop that
+	// arrived before now (raced startup), so the connector context is canceled
+	// immediately in that case.
+	connectorCtx, connectorCtxCancel := n.stopper.start()
+	defer connectorCtxCancel()
 
 	// openMsgTracker tracks open messages until they are acked or nacked
 	var openMsgTracker OpenMessagesTracker
@@ -247,18 +245,7 @@ func (n *SourceNode) stopGraceful(ctx context.Context, reason error) (err error)
 
 func (n *SourceNode) ForceStop(ctx context.Context) {
 	n.logger.Warn(ctx).Msg("force stopping source connector")
-	n.mctx.Lock()
-	defer n.mctx.Unlock()
-	// connectorCtxCancel is nil if Run has not yet reached its initialization
-	// (force-stop raced source startup). Skipping the cancel is safe: stopForceful
-	// kills the pipeline tomb before calling ForceStop, so Run will observe the
-	// canceled node context and stop as soon as it starts. Canceling the connector
-	// context here only interrupts an in-flight Read faster — which by definition
-	// isn't happening yet. Guarding avoids a nil-deref panic (invariant 7: force
-	// at any instant must be recoverable, not crash). See issue #2539.
-	if n.connectorCtxCancel != nil {
-		n.connectorCtxCancel()
-	}
+	n.stopper.stop()
 }
 
 func (n *SourceNode) Pub() <-chan *Message {

--- a/pkg/lifecycle/stream/source_test.go
+++ b/pkg/lifecycle/stream/source_test.go
@@ -221,6 +221,21 @@ func TestSourceNode_StopWhileNextNodeIsStuck(t *testing.T) {
 	is.True(!ok)  // expected node to close outgoing channel
 }
 
+// TestSourceNode_ForceStop_BeforeRun guards #2539: force-stopping a source node
+// before Run has initialized its connector-context cancel must not panic. Before
+// the fix, ForceStop nil-dereferenced connectorCtxCancel and crashed the process
+// (invariant 7: force at any instant must be recoverable, not crash).
+func TestSourceNode_ForceStop_BeforeRun(t *testing.T) {
+	node := &SourceNode{
+		Name:          "source-node",
+		PipelineTimer: noop.Timer{},
+	}
+
+	// connectorCtxCancel is nil here — Run was never called. This must be a no-op,
+	// not a panic.
+	node.ForceStop(context.Background())
+}
+
 func TestSourceNode_ForceStop(t *testing.T) {
 	ctrl := gomock.NewController(t)
 

